magento/magento2
# [Backport-2.2] Code generation improvement for php 7.1

Code generation improvements.
### Description
Improve code generation for proxies and interceptors.
Fix not supported nullable return types and parameters in interceptors, variadics and return types like 'self'.
Make changes compatible with PHP 7.0.
Rework of #16337

### Fixed Issues

1. magento/magento2#15505: Interceptor class methods do not support nullable return types
2. MAGETWO-89899: Proxy/Interceptor generator does not understand PHP 7.1 syntax
3. MAGETWO-95415: Can't generate Plugin if object method returns self

### Manual testing scenarios
Create class with usage:
1. variadics
2. nullable type return/parameter
3. return type 'self'
Check in scope of PHP7.0 and PHP7.1

### Contribution checklist (*)
